### PR TITLE
Boxing+

### PR DESCRIPTION
**Pull request by breadhunt**
## About The Pull Request

Absolutely steals CM's KO audio file and copies their boxing idea.
Punching people with boxing gloves now deals 10 stam damage. Under 0 stamina (5 hits), a KO sound plays and they're knocked out for about 10 seconds.
Numbers can be changed however but this probably won't affect balance too badly. Probably

Also adds handgloves. Wear them, alt-click them while worn, you'll get stuff in your hands. Could be used for some cool shit

Don't merge, still todo icons, KOs, banging the gloves together

## Why It's Good For The Game

We already have boxing areas in some maps and they're pretty fun even without this.
More fluff is always nice and fleshes out the game more
Also I want to punch the fucking FC in the face let me at him let me at him I WANT HIS FUCKING THROAT!!!

There is a risk of this being used in actual combat, but a) a machete is 3x as effective, b) a gun is 10x as effective, c) that would only be applicable in HvH, d) if 2+ people with guns get boxed away, massive skill issue and e) it'd be fucking funny so I don't think it's too much of a problem. If it's too strong at picking off people with low stam (i.e they just finished running) required stamina loss to KO can be reduced to, like, twice as much needed or something

## Changelog
:cl:
soundadd: Steals CM's boxing KO sound. Thanks CM!
add: Boxing gloves now deal stamina damage. Hit someone enough and that will KO them. Still deals brute damage like normal punches, so be careful.
/:cl:

### CODE REVIEW

**Review by BraveMole — CHANGES_REQUESTED** [review]
Lel cm code.

Ping me so I can tell you how to do it properly without snowflake code

**Comment by Andrewhedge** [comment]
if we add funny sound as this on power fist, its usage rate will skyrocket. ' YOU GOT PLANTED!!! '

**Review by Kuro020 — APPROVED** [review]

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 1]
```diff
@@ -69,31 +69,141 @@
 	permeability_coefficient = 0.9
 	siemens_coefficient = 0.9
 
+//Special type of gloves. Alt click and you get some special nodrop gloves
+/obj/item/clothing/gloves/heldgloves
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:73]: you need to make sure both gloves are QDELL_NULL on destroy

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 2]
```diff
@@ -69,31 +69,141 @@
...
+	. = ..()
+	remove_gloves(unequipper)
+
+//We use alt-click to activate/deactive the gloves in-hand
+/obj/item/clothing/gloves/heldgloves/AltClick(mob/user)
+	var/mob/living/carbon/human/wearer = user
+	if(wearer.gloves != src) //We have to be wearing the gloves first
+		return
+
+	if(!remove_gloves(user))
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:92]: that's unintuitive, remove gloves should return true if you can and did remove gloves

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 3]
```diff
@@ -69,31 +69,141 @@
...
+		return
+
+	user.drop_all_held_items() //Gloves require free hands
+	create_gloves(user)
+
+/// Creates the held items for user and puts it in their hand
+/obj/item/clothing/gloves/heldgloves/proc/create_gloves(mob/user)
+	if(user.l_hand != null || user.r_hand != null)
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:100]:
```suggestion
	if(user.l_hand || user.r_hand)
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 4]
```diff
@@ -69,31 +69,141 @@
...
+		return
+
+	var/obj/item/weapon/heldglove/boxing/rightglove = new right_glove
+	user.put_in_r_hand(rightglove, TRUE)
 
+	var/obj/item/weapon/heldglove/boxing/leftglove = new left_glove
+	user.put_in_l_hand(leftglove, TRUE)
+	return
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:108]:
```suggestion
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 5]
```diff
@@ -69,31 +69,141 @@
...
 
-/obj/item/clothing/gloves/boxing
+/// Removes gloves. Returns false if gloves are not currently worn
+/obj/item/clothing/gloves/heldgloves/proc/remove_gloves(mob/user)
+	if(istype(user.l_hand, /obj/item/weapon/heldglove))
+		qdel(user.l_hand)
+		if(istype(user.r_hand, /obj/item/weapon/heldglove))
+			qdel(user.r_hand)
+			return FALSE
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:116]:
```suggestion
		removed = TRUE
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 6]
```diff
@@ -69,31 +69,141 @@
...
+	user.put_in_r_hand(rightglove, TRUE)
 
+	var/obj/item/weapon/heldglove/boxing/leftglove = new left_glove
+	user.put_in_l_hand(leftglove, TRUE)
+	return
 
-/obj/item/clothing/gloves/boxing
+/// Removes gloves. Returns false if gloves are not currently worn
+/obj/item/clothing/gloves/heldgloves/proc/remove_gloves(mob/user)
+	if(istype(user.l_hand, /obj/item/weapon/heldglove))
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:112]:
```suggestion
	var/removed = FALSE
	if(istype(user.l_hand, /obj/item/weapon/heldglove))
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 7]
```diff
@@ -69,31 +69,141 @@
...
+		qdel(user.l_hand)
+		if(istype(user.r_hand, /obj/item/weapon/heldglove))
+			qdel(user.r_hand)
+			return FALSE
+	if(istype(user.r_hand,/obj/item/weapon/heldglove))
+		qdel(user.r_hand)
+		return FALSE
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:119]:
```suggestion
		removed = TRUE
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 8]
```diff
@@ -69,31 +69,141 @@
...
+	return TRUE
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:120]:
```suggestion
	return removed
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 9]
```diff
@@ -69,31 +69,141 @@
...
-/obj/item/clothing/gloves/boxing
+/// Removes gloves. Returns false if gloves are not currently worn
+/obj/item/clothing/gloves/heldgloves/proc/remove_gloves(mob/user)
+	if(istype(user.l_hand, /obj/item/weapon/heldglove))
+		qdel(user.l_hand)
+		if(istype(user.r_hand, /obj/item/weapon/heldglove))
+			qdel(user.r_hand)
+			return FALSE
+	if(istype(user.r_hand,/obj/item/weapon/heldglove))
+		qdel(user.r_hand)
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:118]:
```suggestion
		QDEL_NULL(user.r_hand)
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 10]
```diff
@@ -69,31 +69,141 @@
...
 
+	var/obj/item/weapon/heldglove/boxing/leftglove = new left_glove
+	user.put_in_l_hand(leftglove, TRUE)
+	return
 
-/obj/item/clothing/gloves/boxing
+/// Removes gloves. Returns false if gloves are not currently worn
+/obj/item/clothing/gloves/heldgloves/proc/remove_gloves(mob/user)
+	if(istype(user.l_hand, /obj/item/weapon/heldglove))
+		qdel(user.l_hand)
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:113]:
```suggestion
		QDEL_NULL(user.l_hand)
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 11]
```diff
@@ -69,31 +69,141 @@
...
+		if(istype(user.r_hand, /obj/item/weapon/heldglove))
+			qdel(user.r_hand)
+			return FALSE
+	if(istype(user.r_hand,/obj/item/weapon/heldglove))
+		qdel(user.r_hand)
+		return FALSE
+	return TRUE
+
+/obj/item/weapon/heldglove
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:122]: need a no drop flag

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 12]
```diff
@@ -69,31 +69,141 @@
...
+/obj/item/clothing/gloves/heldgloves/AltClick(mob/user)
+	var/mob/living/carbon/human/wearer = user
+	if(wearer.gloves != src) //We have to be wearing the gloves first
+		return
+
+	if(!remove_gloves(user))
+		return
+
+	user.drop_all_held_items() //Gloves require free hands
+	create_gloves(user)
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:96]: need a no drop flag added and removed on create and remove gloves

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 13]
```diff
@@ -69,31 +69,141 @@
...
+	name = "glove"
+
+
+//Boxing gloves
+/obj/item/clothing/gloves/heldgloves/boxing
 	name = "boxing gloves"
 	desc = "Because you really needed another excuse to punch your crewmates."
 	icon_state = "boxing"
 	item_state = "boxing"
+	left_glove = /obj/item/weapon/heldglove/boxing/hook
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:132]: no, either it is a object or a path. You are using it as an object, so don't do that
```suggestion
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 14]
```diff
@@ -69,31 +69,141 @@
...
+	right_glove = /obj/item/weapon/heldglove/boxing/jab
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:133]:
```suggestion
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 15]
```diff
@@ -69,31 +69,141 @@
...
+
+/obj/item/weapon/heldglove/boxing/jab
+	icon_state = "boxing_j"
+	attack_verb = list("jabbed")
+
+
+/obj/item/clothing/gloves/heldgloves/boxing/green
 	icon_state = "boxinggreen"
 	item_state = "boxinggreen"
+	left_glove = /obj/item/weapon/heldglove/boxing/hook/green
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:166]: ditto

**Review by BraveMole — CHANGES_REQUESTED** [review]
Object reference != path.

Your left and right gloves should be object references, so you can QDEL the object when needed.

I suggest adding another var, calling it gloves_path, and doing a : 

right_glove = new gloves_path()
left_glove = new gloves_path() when needed

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 16]
```diff
@@ -69,31 +69,147 @@
 	permeability_coefficient = 0.9
 	siemens_coefficient = 0.9
 
+//Special type of gloves. Alt click and you get some special nodrop gloves
+/obj/item/clothing/gloves/heldgloves
+	name = "gloves"
+	var/rightglove_path
+	var/leftglove_path
+	var/obj/item/weapon/heldglove/rightglove
+	var/obj/item/weapon/heldglove/leftglove
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:78]: you need to autodoc everything

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 17]
```diff
@@ -69,31 +69,147 @@
...
+	user.drop_all_held_items() //Gloves require free hands
+	create_gloves(user)
 
+/// Creates the held items for user and puts it in their hand
+/obj/item/clothing/gloves/heldgloves/proc/create_gloves(mob/user)
+	if(user.l_hand || user.r_hand)
+		return
+
+	rightglove = new rightglove_path()
+	TOGGLE_BITFIELD(rightglove.flags_item, NODROP)
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:111]: why are they not no drop in the define ?

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 18]
```diff
@@ -69,31 +69,147 @@
 	permeability_coefficient = 0.9
 	siemens_coefficient = 0.9
 
+//Special type of gloves. Alt click and you get some special nodrop gloves
+/obj/item/clothing/gloves/heldgloves
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:73]: this needs to be no drop

File: code/modules/clothing/gloves/miscellaneous.dm (hunk 12 above)
Comment (BraveMole) [reply code/modules/clothing/gloves/miscellaneous.dm:103]: the no drop should be added on the clothing item, and be present by default on the glove held one

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 19]
```diff
@@ -69,31 +69,151 @@
...
+	QDEL_NULL(leftglove)
+	return ..()
+
+/obj/item/clothing/gloves/heldgloves/examine(mob/user)
+	. = ..()
+	to_chat(user, "<span class='notice'>Alt-click the gloves when worn to strap them into your hands.")
+
+/obj/item/clothing/gloves/heldgloves/unequipped(mob/unequipper, slot)
+	. = ..()
+	remove_gloves(unequipper)
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:95]: no longer usefull

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 20]
```diff
@@ -69,31 +69,151 @@
...
+
+//We use alt-click to activate/deactive the gloves in-hand
+/obj/item/clothing/gloves/heldgloves/AltClick(mob/user)
+	var/mob/living/carbon/human/wearer = user
+	if(wearer.gloves != src) //We have to be wearing the gloves first
+		return
+
+	if(remove_gloves(user))
+		return
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:104]: remove the no drop bit flag here

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 21]
```diff
@@ -69,31 +69,151 @@
...
 
+	user.drop_all_held_items() //Gloves require free hands
+	create_gloves(user)
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:107]: add the no drop flag here

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 22]
```diff
@@ -69,31 +69,151 @@
...
+/obj/item/clothing/gloves/heldgloves/proc/remove_gloves(mob/user)
+	var/removed = FALSE
+	if(leftglove)
+		QDEL_NULL(leftglove)
+		removed = TRUE
+	if(rightglove)
+		QDEL_NULL(rightglove)
+		removed = TRUE
+	if(removed)
+		TOGGLE_BITFIELD(flags_item, NODROP)
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:130]: don't add it here, and no toggle, just enable or disalbe

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/clothing/gloves/miscellaneous.dm (hunk 19 above)
Comment (breadhunt) [reply code/modules/clothing/gloves/miscellaneous.dm:95]: Still called on delimbs, clears up the glove nodrop as well as the weapon gloves themselves. They'd just plop on the floor without this

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 23]
```diff
@@ -69,31 +69,162 @@
...
+	icon_state = "boxing"
+	item_state = "boxing"
+	damtype = STAMINA
+	force = 10
+	w_class = WEIGHT_CLASS_BULKY
+	hitsound = "punch"
+
+/obj/item/weapon/heldglove/boxing/attack(mob/living/M, mob/living/user)
+	. = ..()
+	if(ishuman(M))
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:168]: early return this

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 24]
```diff
@@ -69,31 +69,162 @@
...
+		var/mob/living/carbon/human/target = M
+		if(target.getStaminaLoss() > 10 && (target.stat != UNCONSCIOUS))
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:170]: early return here as well

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 25]
```diff
@@ -69,31 +69,162 @@
...
+
+/obj/item/clothing/gloves/heldgloves/boxing/green
 	icon_state = "boxinggreen"
 	item_state = "boxinggreen"
+	rightglove_path = /obj/item/weapon/heldglove/boxing/hook/green
+	leftglove_path = /obj/item/weapon/heldglove/boxing/jab/green
+
+/obj/item/weapon/heldglove/boxing/hook/green
+	icon_state = "boxing_p_g"
+/obj/item/weapon/heldglove/boxing/jab/green
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:192]:
```suggestion

/obj/item/weapon/heldglove/boxing/jab/green
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 26]
```diff
@@ -69,31 +69,162 @@
...
-/obj/item/clothing/gloves/boxing/blue
+/obj/item/clothing/gloves/heldgloves/boxing/blue
 	icon_state = "boxingblue"
 	item_state = "boxingblue"
+	rightglove_path = /obj/item/weapon/heldglove/boxing/hook/blue
+	leftglove_path = /obj/item/weapon/heldglove/boxing/jab/blue
+
+/obj/item/weapon/heldglove/boxing/hook/blue
+	icon_state = "boxing_p_b"
+/obj/item/weapon/heldglove/boxing/jab/blue
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:203]:
```suggestion

/obj/item/weapon/heldglove/boxing/jab/blue
```

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 27]
```diff
@@ -69,31 +69,162 @@
...
-/obj/item/clothing/gloves/boxing/yellow
+/obj/item/clothing/gloves/heldgloves/boxing/yellow
 	icon_state = "boxingyellow"
 	item_state = "boxingyellow"
+	rightglove_path = /obj/item/weapon/heldglove/boxing/hook/yellow
+	leftglove_path = /obj/item/weapon/heldglove/boxing/jab/yellow
+
+/obj/item/weapon/heldglove/boxing/hook/yellow
+	icon_state = "boxing_p_y"
+/obj/item/weapon/heldglove/boxing/jab/yellow
```
Comment (BraveMole) [inline code/modules/clothing/gloves/miscellaneous.dm:214]:
```suggestion

/obj/item/weapon/heldglove/boxing/jab/yellow
```

**Review by BraveMole — APPROVED** [review]

File: code/modules/clothing/gloves/miscellaneous.dm [hunk 28]
```diff
@@ -69,31 +69,169 @@
...
+	var/obj/item/weapon/heldglove/leftglove
+
+/obj/item/clothing/gloves/heldgloves/Destroy()
+	QDEL_NULL(rightglove)
+	QDEL_NULL(leftglove)
+	return ..()
+
+/obj/item/clothing/gloves/heldgloves/examine(mob/user)
+	. = ..()
+	to_chat(user, "<span class='notice'>Alt-click the gloves when worn to strap them into your hands.")
```
Comment (TiviPlus) [inline code/modules/clothing/gloves/miscellaneous.dm:91]:
```suggestion
	. = ..()
	. += "Alt-click the gloves when worn to strap them into your hands."
```

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Pull request merged**